packages/cmf/src/sagaRouter/router.js

```diff
@@ -30,31 +30,55 @@ function shouldRestartSaga(maybeSaga, match) {
 	return false;
 }
 
+function parseSagaState(routeFragments, sagas, currentLocation, index) {
+	const routeFragment = routeFragments[index];
+	return {
+		routeFragment,
+		match: matchPath(currentLocation.pathname, { path: routeFragment }),
+		maybeSaga: sagas[routeFragment],
+	};
+}
+
 export default function* routerSaga(history, routes) {
 	const sagas = {};
+	let shouldStart = [];
+	const routeFragments = Object.keys(routes);
 	while (true) {
 		yield take('@@router/LOCATION_CHANGE');
+		shouldStart = [];
 		const currentLocation = history.getCurrentLocation();
-		const routeFragments = Object.keys(routes);
 		for (let index = 0; index < routeFragments.length;) {
-			const routeFragment = routeFragments[index];
-			const match = matchPath(currentLocation.pathname, { path: routeFragment });
-			const maybeSaga = sagas[routeFragment];
-			if (shouldStartSaga(maybeSaga, match)) {
-				sagas[routeFragment] = {
-					saga: yield spawn(routes[routeFragment], match.params),
-					match,
-				};
-			} else if (shouldCancelSaga(maybeSaga, match)) {
+			const { routeFragment, match, maybeSaga } = parseSagaState(
+				routeFragments,
+				sagas,
+				currentLocation,
+				index,
+			);
+			if (shouldCancelSaga(maybeSaga, match)) {
 				yield cancel(maybeSaga.saga);
 			} else if (shouldRestartSaga(maybeSaga, match)) {
 				yield cancel(maybeSaga.saga);
 				sagas[routeFragment] = {
 					saga: yield spawn(routes[routeFragment], match.params),
 					match,
 				};
+			} else if (shouldStartSaga(maybeSaga, match)) {
+				shouldStart.push(routeFragment);
 			}
 			index += 1;
 		}
+		for (let index = 0; index < shouldStart.length;) {
+			const { routeFragment, match } = parseSagaState(
```

Comment:
Can you store the result of parseSagaState() in shouldStart instead of calling it again ?